File: linear_search.py
# implementation of linear search using recursion
# Hlulani Myambo
# 20 May 2025
# Time-complexity -> O(n)

def linear_search(array: list[int], target: int, length: int, index=0) -> int:
    # empty array
    if not array:
        return -1
    if array[index] == target:
        return index
    # out of bound
    if index >= length:
        return -1
    return linear_search(array, target, length, index + 1)

test_case: list[int] = [10, 78, 12, 90, 91, 50, 21, -1, 921, 200, 1, 0, 59, 96, 1000]
length: int = len(test_case) - 1
target: int = 21

print(linear_search(test_case, target, length)) #ouput = 6
print(linear_search(test_case, 20, length)) #ouput = -1
print(linear_search(test_case, 10, length)) #ouput = 0
print(linear_search(test_case, 1000, length)) #output = 14

File: LinearSearch.java
// Linear Search using recursion
// Myambo Hlulani
// 22 May 2025
// => time-complexity -> O(n)
// => as input increases, time to process it increases

public class LinearSearch {
    public static int search(int[] array, int target, int index) {
        if (array.length == 0) {
            return -1;
        }

        if (array.length - 1 >= index) {
            if (array[index] == target) {
                return index;
            }
            return search(array, target, index + 1);
        }

        return -1;
    }

    public static void main(String[] args) {
        int[] ARRAY = { 10, 20, 30, 40, 50 };
        int target = 10;
        int INDEX = 0;

        // Test cases
        System.out.println(search(ARRAY, target, INDEX)); // output = 0
        System.out.println(search(ARRAY, 0, INDEX)); // output = -1
        System.out.println(search(ARRAY, 30, INDEX)); // output = 2
        System.out.println(search(ARRAY, 50, INDEX)); // output = 4
        System.out.println(search(ARRAY, 60, INDEX)); // output = -1
    }
}